File: problem5.py
"""
Project Euler Problem 5

EN:2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

TUR: 2520 1 den 10 a kadar tüm sayıların kalansız bölündüğü en küçük sayıdır

1'den 20'ye kadar tüm sayıların kalansız böldüğü en küçük sayı nedir?

"""

# Problemi anlamak: Problem bizden 1 den 20'ye kadarki tüm sayıların ekokunu istiyor
# Defterimizde tablo oluştururcasına çözeceğiz bu soruyu
    
# 1 den 20'ye tüm asal sayılar
# Bu asal sayıları EKOK hesaplarken kullanacağız
all_prime_factors = [2,3,5,7,11,13,17,19]
    
numbers = list(range(1,21))

# Kullandığımız çarpanları bu listeye ekleyeceğiz
used_factors = []

"""
Hatırlıyorsanız EKOK hesaplarken tüm sayıları 1 yapana dek asal bölüyorduk
yine öyle yapacağız bu yüzden tüm sayılar bir olana dek çalışacak bir döngü başlattık.
"""
while sum(numbers) != 20:
    
    # Asal çarpanlarımızı tek tek döndürüyoruz
    for factor in all_prime_factors:
        check = False
        # 1 den 20 ye sayılarımızı tek tek döndürüyoruz
        for number in numbers:
            # Eğer sayı asal sayımıza bölünüyorsa bölüyoruz ve bu asal sayıyı
            # kullandığımızı tespit etmek için False'u True'ya dönüştürüyoruz
            if number%factor == 0:
                check = True
                numbers[numbers.index(number)] = number//factor
        # Eğer sayıyı kullandıysak 
        # kullanılan çarpanlar listesine ekliyoruz
        if check:
            used_factors.append(factor)
                
        

    
result = 1
for i in used_factors:
    result = result*i
    
print(result)
            

File: problem5.cs
using System;
using System.Collections.Generic;

namespace problem5
{
    class Program
    {
        static void Main(string[] args)
        {
            /*
             Project Euler Problem 5
                
            EN:2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
            What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
             
            TUR: 2520 1 den 10 a kadar tüm sayıların kalansız bölündüğü en küçük sayıdır
            1'den 20'ye kadar tüm sayıların kalansız böldüğü en küçük sayı nedir?
            */
            int ekok=1;
            int ekokBul(int sayi1, int sayi2)
            {
                //ekok = sayıların çarpımı / ebob
                int _ebob = 1, _ekok = 1;
                //ebob bulmak için sayılarımızdan büyük olanını alıp bir bir azaltıyoruz ve her iki sayıyı da bölebilen bir sayı arıyoruz.
                for (int i = Math.Max(sayi1, sayi2); i > 1; i--)
                {
                    //sayıyı bulduğumuzda
                    if (sayi1 % i == 0 && sayi2 %i == 0)
                    {
                        //sayıyı eboba eşitliyoruz
                        _ebob = i;
                        //Çarpım int değerlerini aşabileceğinden eğer ebob sayılardan küçük olana eşitse ekok büyük olandır diye bir işlem yapıyoruz. Altta daha detaylı...
                        if (_ebob == Math.Min(sayi1,sayi2))
                        {
                            return Math.Max(sayi1,sayi2);
                        }
                        break;
                    }
                }
                //Eğer ebob sayılardan küçük olana eşit değilse döngü bitirilir ve sayıların çarpımı / ebob formülünden ekok bulunur
                _ekok = ((sayi1 * sayi2) / _ebob);
                return _ekok;
            }
            //Sırasıyla 2'den 20'ye kadar olan sayıları alıp ekoklarını buluyoruz. Aşağıda daha detaylı...
            for (int i = 2; i < 21; i++)
            {
                ekok = ekokBul(i,ekok);
            }
            Console.WriteLine("sonuç: " + ekok);
        }    
    }
}
/*
 Ebob ekok için açıklama:
    Eğer iki sayının en büyük ortak böleni yani ebobu küçük olanına eşitse iki sayı büyük olan sayıda birleşir.
    Örneğin: 2 ve 4 sayıları küçük olan ebob (İki sayımız da 2'ye bölünebilir.) bbu sayılar ise 4 sayısında birleşir yani ortak katı 4'tür.
    Not: Bunu yapmamın amacı çarpım durumunda int veri tipinin değer aralığından daha fazla değer almaya çalışmasıdır.

 Ekok için açıklama:
    Sırasıyla sayıları ve ekokları buluyoruz işlem şöyle işliyor sayımız 2 ekokumuzun ilk değeri 1 ise 2 ve 1'in ekoku 2'dir. Bunu bulup ekok değişkenimize atarız ve sıra 3'e gelir.
    Bu sefer 3 ve 2'nin (önceki ekok değerimiz) ekokunu buluruz bu da 6 yapar ekok değişkenini 6 yaparız ve sıra 4'e geçer. 4 ile önceki ekok değerimiz olan 6'nın ekokunu buluruz... 
    İşlemler böyle devam eder gider.
*/
